src/main/host/descriptor/mod.rs

```diff
@@ -241,35 +254,48 @@ impl StateEventSource {
         }
     }
 
+    /// Add a listener. The filter applies only to state changes, not signals. Only signals that are
+    /// monitored will be passed to the listener.
     pub fn add_listener(
         &mut self,
-        monitoring: FileState,
+        monitoring_state: FileState,
+        monitoring_signals: FileSignals,
         filter: StateListenerFilter,
-        notify_fn: impl Fn(FileState, FileState, &mut CallbackQueue) + Send + Sync + 'static,
-    ) -> Handle<(FileState, FileState)> {
-        self.inner.add_listener(move |(state, changed), cb_queue| {
-            // true if any of the bits we're monitoring have changed
-            let flipped = monitoring.intersects(changed);
-
-            // true if any of the bits we're monitoring are set
-            let on = monitoring.intersects(state);
-
-            let notify = match filter {
-                // at least one monitored bit is on, and at least one has changed
-                StateListenerFilter::OffToOn => flipped && on,
-                // all monitored bits are off, and at least one has changed
-                StateListenerFilter::OnToOff => flipped && !on,
-                // at least one monitored bit has changed
-                StateListenerFilter::Always => flipped,
-                StateListenerFilter::Never => false,
-            };
-
-            if !notify {
-                return;
-            }
-
-            (notify_fn)(state, changed, cb_queue)
-        })
+        notify_fn: impl Fn(FileState, FileState, FileSignals, &mut CallbackQueue)
+            + Send
+            + Sync
+            + 'static,
+    ) -> StateListenHandle {
+        self.inner
+            .add_listener(move |(state, changed, signals), cb_queue| {
+                // true if any of the bits we're monitoring have changed
+                let flipped = monitoring_state.intersects(changed);
+
+                // true if any of the bits we're monitoring are set
+                let on = monitoring_state.intersects(state);
+
+                let notify = match filter {
+                    // at least one monitored bit is on, and at least one has changed
+                    StateListenerFilter::OffToOn => flipped && on,
+                    // all monitored bits are off, and at least one has changed
+                    StateListenerFilter::OnToOff => flipped && !on,
+                    // at least one monitored bit has changed
+                    StateListenerFilter::Always => flipped,
+                    StateListenerFilter::Never => false,
+                };
```

Comment:
I wonder if we could have just extended the `StateListenerFilter` enum to have e.g. a `StateListenerFilter::OffToOff` and a `StateListenerFilter::OnToOn` enum variants rather than separating signals into a separate construct. But perhaps separating the signals as you have done is more flexible and/or easier to understand in the code. I'm still fine merging as-is, but wanted to bring this up in case you hadn't considered it before but think it might be a simpler design to deal with in the code going forward.

Reply:
My reason against this was that it doesn't tell you anything about why the listener was notified. If the listener was called and given a state change of "nothing changed", you still wouldn't know why you were notified in the first place. You could interpret some meaning from it, but then you could accidentally add a seemingly inconsequential state change in a socket that then gets interpreted incorrectly. For example something like:

```rust
// should already be readable, but set the readable state flag here just to be sure
self.update_state(FileState::READABLE, true);
```

But now we'd be sending an "OnToOn" file state event, which might get interpreted as having some certain meaning in a listener, but the calling code didn't intend this.

Also the event source doesn't have a way to set an individual flag. So if you were only monitoring the "READABLE" flag with the "OffToOff" filter, but then the file changes only its "WRITABLE" flag, the listener watching the "READABLE" flag will still run even though the update has nothing to do with the "READABLE" flag since the state was technically updated.

Reply:
Ok, makes sense to me - thanks for thinking this through!